core/sync/dependency_graph.js

```diff
@@ -0,0 +1,236 @@
+ * C to the graph as it would create a cycle (i.e. we compare A and C before
+ * comparing B and C so the dependency betwee C and A is preserved).
+ */
+
+/*::
+import type { Change } from './'
+
+type GraphOptions = {
+  compare: (Change, Change) => number
+}
+type Node = {
+  change: Change,
+  dependentOn: Node[],
+};
+*/
+
+class DependencyGraph {
+  /*::
+  nodes: Node[]
+  compare: (Change, Change) => number
+  */
+
+  // The compare method is expected to be the kind of method that would be
+  // passed to `Array.sort()`. It should return:
+  // - a negative value when its first Change argument should be propagated
+  //   before its second Change argument
+  // - a positive value when the second Change should be propagated before the
+  //   first one
+  // - 0 when either Change can be propagated first
+  constructor(changes /*: Change[] */, { compare } /*: GraphOptions */) {
+    this.compare = compare
+    this.nodes = []
+
+    for (const change of changes) {
+      this.insert(change)
+    }
+  }
+
+  // Create a new Node from the given change and compute all the edges between
+  // this Node and the other Nodes in the graph by comparing changes together.
+  insert(change /*: Change */) {
+    const newNode = {
+      change,
+      dependentOn: []
+    }
+
+    for (const node of this.nodes) {
+      const order = this.compare(node.change, newNode.change)
+
+      if (order < 0) {
+        // node should come first
+        newNode.dependentOn.push(node)
+
+        if (createsCycle(newNode)) {
+          // We can't deal with circular dependencies for now so we simply avoid
+          // them altogether.
+          // This means we might end up with dependent changes being
+          // synchronized before their dependencies but at least we won't end up
+          // in an infinite loop when expanding the graph into a sorted array.
+          newNode.dependentOn.splice(-1)
+        }
+      } else if (order > 0) {
+        // newNode should come first
+        node.dependentOn.push(newNode)
+
+        if (createsCycle(newNode)) {
+          // We can't deal with circular dependencies for now so we simply avoid
+          // them altogether.
+          // This means we might end up with dependent changes being
+          // synchronized before their dependencies but at least we won't end up
+          // in an infinite loop when expanding the graph into a sorted array.
+          node.dependentOn.splice(-1)
+        }
+      } else {
+        // No dependencies between node and newNode
+      }
+    }
+
+    this.nodes.push(newNode)
+  }
+
+  // Expand the graph into an ordered list of Node objects and then map this
+  // list to a list of Change objects which can be used by Sync to propagate
+  // them.
+  toArray() /*: Change[] */ {
+    const orderedChanges = new Set()
+
+    for (const node of this.nodes) {
+      for (const dependency of orderedDependencies(node)) {
```

Comment:
What are the performances expectation? The algorithm has a pretty bad "worst case" with long chain of dependencies. Maybe we don't care, but I would like your opinion on that.

Reply:
I haven't tried to propagate a large number of changes (I will though) but I don't expect great performances as the complexity of the algorithm isn't so good.
I don't expect our users to have lots of dependencies between their changes though and if it should happen, I prefer to make sure we can handle them even if it takes some time to sort them out (I don't expect the expansion to take more time than the synchronization itself).

Reply:
I have revisited the expansion algorithm to make sure each node is visited only once and remove a lot of loops that were unnecessary.
This makes the algorithm way easier to understand as well.